### PR TITLE
Added segment about string types and their corresponding member functions

### PR DESCRIPTION
**Pull request by Nizarll**
i think this segment was pretty important especially since the programming language targets people from a c background so providing an explanation of the underlying implementations .

### CODE REVIEW

**Comment by Nizarll** [comment]
However i'm thinking about making a different segment for member functions of most of the underlying types something similar to cppreference where any type from the standard library or the base types that states all of their member functions shows examples and usage

**Comment by joshring** [comment]
> However i'm thinking about making a different segment for member functions of most of the underlying types something similar to cppreference where any type from the standard library or the base types that states all of their member functions shows examples and usage

A separate section might be good (second PR?), people have also asked about the mathematical operations defined as member functions too.

File: src/content/docs/Language Common/strings.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+#### ZString
+
+```c3
+
+distinct ZString = inline char*;
+```
+
+\
+A ZString is a distinct type for null-terminated C-style strings ` char* `. Converting a ` char* ` to a ` ZString ` requires an explicit cast to ensure the safety and `\0` termination of the string data.
```
Comment (joshring) [inline src/content/docs/Language Common/strings.md]: Perhaps we can add the use-case of this string type:

```suggestion
ZString is used when working with C code, which expects null-terminated C-style strings of type `char*`. it is a distinct type so converting to a `ZString` requires an explicit cast. This helps to remind the user to check there is appropriate `\0` termination of the string data. 

The [ZString methods](#zstring-member-functions) are outlined below.
```

File: src/content/docs/Language Common/strings.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+#### String
+
+```c3
+
+distinct String = inline char[];
+
+```
+\
+Strings are usually the typical type to use, they can be sliced , compared etc ... \
```
Comment (joshring) [inline src/content/docs/Language Common/strings.md:17]: Strings are the typical string type to use, they can be sliced, compared, split etc, the [string methods](#string-member-functions) are outlined below.
It is possible to access the length of a `String` instance through the  `.len`  member, eg
```c3
String my_string = "example";
int length = my_string.len
```

File: src/content/docs/Language Common/strings.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+fn WString! String.to_new_wstring(s) => s.to_wstring(allocator::heap());
+```
+```c3
+fn WString! String.to_temp_wstring(s) => s.to_wstring(allocator::temp());
+```
+```c3
+fn StringIterator String.iterator(s);
+```
+
+#### ZString :
```
Comment (joshring) [inline src/content/docs/Language Common/strings.md]: So we can link directly to this heading (H3 is linkable, H4 is not)
```suggestion
### ZString Member Functions
```

File: src/content/docs/Language Common/strings.md [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+distinct DString (OutStream) = void*;
+```
+
+\
+`DString` is a dynamic string builder that supports various string operations at runtime, allowing for flexible manipulation without the need for manual memory allocation.
+
+## Member functions:
+
+#### String :
```
Comment (joshring) [inline src/content/docs/Language Common/strings.md]: So we can link directly to this heading (H3 is linkable, H4 is not)
```suggestion
### String Member Functions
```

**Review by joshring — CHANGES_REQUESTED** [review]
Thanks for this!
Some small bits, added some suggestions to speed things along :+1:

**Comment by Nizarll** [comment]
> > However i'm thinking about making a different segment for member functions of most of the underlying types something similar to cppreference where any type from the standard library or the base types that states all of their member functions shows examples and usage
> 
> A separate section might be good (second PR?), people have also asked about the mathematical operations defined as member functions too.

i think the first part without the member functions should be kept , but the member functions should probably be in a whole other **Section** about the standard library's type and the functions as well as examples similar to cppref i will be preparing a second pr this week ! I am still wondering about the name of the section though

**Comment by joshring** [comment]
What jumps to mind immediately is "Type Member Functions"

File: src/content/docs/Language Common/strings.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+distinct String = inline char[];
+
+```
+\
+Strings are usually the typical type to use, they can be sliced , compared etc ... \
+It is possible to access the length of a `String` instance through the  ` .len  `  operator.
+
+
+#### ZString
```
Comment (joshring) [inline src/content/docs/Language Common/strings.md]:
```suggestion
### ZString
```
so we can link, H3 is linkable

File: src/content/docs/Language Common/strings.md [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+```c3
+
+distinct ZString = inline char*;
+```
+
+\
+A ZString is a distinct type for null-terminated C-style strings ` char* `. Converting a ` char* ` to a ` ZString ` requires an explicit cast to ensure the safety and `\0` termination of the string data.
+
+>  **⚠️**  Converting a `String` to a `ZString` should be done carefully to ensure \0 termination of the underlying `char*`.
```
Comment (joshring) [inline src/content/docs/Language Common/strings.md]:
```suggestion
:::caution  
Ensure the terminal `\0` when converting from `String` to `ZString`.
:::
```

File: src/content/docs/Language Common/strings.md [hunk 7]
```diff
@@ -0,0 +1,235 @@
+---
+title: Strings
+description: Strings
+sidebar:
+    order: 62
+---
+In C3, multiple string types are available, each suited to different use cases.
+
+#### String
```
Comment (joshring) [inline src/content/docs/Language Common/strings.md]:
```suggestion
### String
```
H3 is linkable

**Review by joshring — CHANGES_REQUESTED** [review]

**Comment by joshring** [comment]
Thanks @Nizarll could you apply those above? then we can look to get this merged :)

**Comment by Nizarll** [comment]
Alright , done.
I've been preparing the pr for default types member functions it should be done in a few days !

**Comment by joshring** [comment]
Still a few suggestions left, but nearly there

**Comment by lerno** [comment]
Is this ok now @joshring

**Review by joshring — APPROVED** [review]

**Comment by joshring** [comment]
Thanks!, all good

**Comment by lerno** [comment]
Thank you for the hard work!

**Pull request merged**